tensorflow/notebooks/cvae.ipynb

```diff
@@ -0,0 +1,1181 @@
+    "            ]\n",
+    "        )\n",
+    "\n",
+    "    def decode(self, z, apply_sigmoid=False):\n",
+    "        logits = self.generative_net.predict(z)\n",
+    "        if apply_sigmoid:\n",
+    "            probs = 1/(1 + np.exp(-logits))\n",
+    "\n",
+    "            return probs\n",
+    "\n",
+    "        return logits"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Get the CVAE model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "latent_dim = 50\n",
+    "vae = CVAE(latent_dim)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "0FMYgY_mPfTi"
+   },
+   "source": [
+    "## Define the loss function and the optimizer\n",
+    "\n",
+    "VAEs train by maximizing the evidence lower bound (ELBO) on the marginal log-likelihood:\n",
+    "\n",
+    "$$\\log p(x) \\ge \\text{ELBO} = \\mathbb{E}_{q(z|x)}\\left[\\log \\frac{p(x, z)}{q(z|x)}\\right].$$\n",
+    "\n",
+    "In practice, we optimize the single sample Monte Carlo estimate of this expectation:\n",
+    "\n",
+    "$$\\log p(x| z) + \\log p(z) - \\log q(z|x),$$\n",
+    "where $z$ is sampled from $q(z|x)$.\n",
+    "\n",
+    "**Note**: we could also analytically compute the KL term, but here we incorporate all three terms in the Monte Carlo estimator for simplicity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {
+    "colab": {
+     "autoexec": {
+      "startup": false,
+      "wait_interval": 0.0
+     }
+    },
+    "colab_type": "code",
+    "id": "iWCn_PVdEJZ7"
+   },
+   "outputs": [],
+   "source": [
+    "def log_normal_pdf(sample, mean, logvar, raxis=1):\n",
+    "    log2pi = tf.log(2. * np.pi)\n",
+    "    return tf.reduce_sum(\n",
+    "        -.5 * ((sample - mean) ** 2. * tf.exp(-logvar) + logvar + log2pi),\n",
+    "        axis=raxis)\n",
+    "\n",
+    "def compute_loss(y_true, y_pred):\n",
+    "    mean, logvar = vae.split_layer(vae.inference_net(y_true))\n",
+    "    z = vae.reparam_layer((mean, logvar))\n",
+    "    cross_ent = tf.nn.sigmoid_cross_entropy_with_logits(logits=y_pred,\n",
+    "                                                        labels=y_true)\n",
+    "    logpx_z = -tf.reduce_sum(cross_ent, axis=[1, 2, 3])\n",
+    "    logpz = log_normal_pdf(z, 0., 0.)\n",
+    "    logqz_x = log_normal_pdf(z, mean, logvar)\n",
+    "    return -tf.reduce_mean(logpx_z + logpz - logqz_x)\n",
+    "    \n",
+    "optimizer = tf.train.AdamOptimizer(5e-3)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Compile the keras model and wrap it with `KerasModel` for distributed training"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "vae.model.compile(optimizer, loss=compute_loss)\n",
```

Comment:
not sure `loss=compute` is correct. see the official Keras implementation https://keras.io/examples/variational_autoencoder/

They use the `add_loss` api